### ISSUE TITLE
Lifecycled: Frequent crashes causes lifecycle hook to not be completed

### ISSUE THREAD

**Issue by itsdalmo** [telia-oss/terraform-aws-concourse#6]
Using latest version of lifecycled:
https://github.com/buildkite/lifecycled/releases/tag/v2.0.1

The process seems to be crashing frequently (every ~1h) with the following error message:
```
Aug 17 09:47:54 ip-10-9-56-15.eu-west-1.compute.internal lifecycled[4574]: time="2018-08-17T09:47:54Z" level=fatal msg="Failed to delete queue: RequestError: send request failed\ncaused by: Post https://sns.eu-west-1.amazonaws.com/: dial tcp: lookup sns.eu-west-1.amazonaws.com on 10.9.0.2:53: no such host"
```

When handling an actual lifecycle hook it never actually completed the lifecycle action, presumably because it crashes with the error above before it is able to send the completion. Worth noting that lifecycled never actually logs that the [handler has completed](https://github.com/buildkite/lifecycled/blob/master/daemon.go#L141) before crashing with the error above. Which suggests that it is trying to delete the queue while running the handler.

**Comment by itsdalmo** [comment]
Some more details:
- The error seems to originate here: https://github.com/buildkite/lifecycled/blob/master/main.go#L79
- Either by being invoked from:
  - A deferred call here: https://github.com/buildkite/lifecycled/blob/master/main.go#L83
  - An OS Signal: https://github.com/buildkite/lifecycled/blob/master/main.go#L96
- The deferred call should not be run before [`daemon.Start()`](https://github.com/buildkite/lifecycled/blob/master/main.go#L108) returns.
- `daemon.Start()` does not contain any blocking calls, except for returning `d.queue.Receive()` which seems to never actually return since it is [continuously polling for new SQS messages](https://github.com/buildkite/lifecycled/blob/master/queue.go#L115) - which to me suggests that `daemon.Start()` will never return either.

The only way I can see lifecycled shutting down, is after completing the lifecycle hook and getting a SIGTERM which [prompts it to delete the queue](https://github.com/buildkite/lifecycled/blob/master/main.go#L96), at which point the `d.queue.Receive()` will [return an error](https://github.com/buildkite/lifecycled/blob/master/queue.go#L108) because the queue does not exist, which is then passed on by `daemon.Start()` and then probably causes the binary to exit with a non-zero exit code.

The question then becomes, why is it trying to delete the queue (and failing) before it has completed the lifecycle hook? Could it somehow be caused by the handler script?:

- lifecycled service: https://github.com/telia-oss/terraform-aws-concourse/blob/master/modules/worker/cloud-config.yml#L121
- handler script: https://github.com/telia-oss/terraform-aws-concourse/blob/master/modules/worker/cloud-config.yml#L151
- concourse shutdown routine: https://github.com/telia-oss/terraform-aws-concourse/blob/master/modules/worker/cloud-config.yml#L69

Could `/bin/kill -s TERM $MAINPID` somehow translate $MAINPID to the PID of lifecycled because that is the process trying to shut down Concourse? Seems strange according to the systemd [documentation](https://www.freedesktop.org/software/systemd/man/systemd.service.html#ExecStop=).

**Comment by itsdalmo** [comment]
This issue should be resolved with the [upgrade to `lifeycled` v2.1.1](https://github.com/telia-oss/terraform-aws-concourse/commit/1966314051bf5878d7d844b71bc3af047064509f).
